### PR TITLE
Ofi 2.0 changes

### PR DESCRIPTION
**Pull request by j-xiong**
Here are the proposed changes for OFI 2.0, most of which are cherry-picked from #9384 and modified to fit current tip of main branch.  Since the sockets and usnic provider have not been removed as suggested by #9384, extra work is needed to keep them in sync with the changes.

More commits are going to be added to this PR. The merging of this PR will embark the forking of `v1.x-main`.

### CODE REVIEW

File: man/fi_domain.3.md [hunk 1]
```diff
@@ -348,12 +352,11 @@ are defined.
   received operations.
 
 *FI_PROGRESS_CONTROL_UNIFIED*
-: This progress model indicates that the user will synchronize progressing the
-  data and control operations themselves (i.e. this allows the control interface
-  to NOT be thread safe). It is only valid for control progress (not data progress).
-  Setting control=FI_PROGRESS_CONTROL_UNIFIED, data=FI_PROGRESS_MANUAL, and
-  threading=FI_THREAD_DOMAIN/FI_THREAD_COMPLETION allows Libfabric to remove all
-  locking in the critical data progress path.
+: This progress model indicates that the user will synchronize progressing
```
Comment (shijin-aws) [inline man/fi_domain.3.md:348]: If we remove control progress, aren't we assuming data and ctrl progress are unified?

Comment (j-xiong) [reply man/fi_domain.3.md:350]: That is a question I would like to discuss. It is possible that both use the same model (auto or manual) but progress independently. But does it make practical sense to keep them separated? Or are there technical difficulties for all providers to support unified progress?

Comment (shijin-aws) [reply man/fi_domain.3.md:350]: cc @shefty

Comment (shefty) [reply man/fi_domain.3.md:350]: My intent was to remove the arbitrary separation between data and control progress.  There's only progress, which can be manual or auto.  Heck, even trying to determine if an async operation that generates a completion meets the requirement for being manual versus auto is non-trivial.

The only "non-data" asynchronous operation that is left is connection establishment.  Such operations are reported to an EQ, which already provides separation from data transfer completions.  Connections require user intervention in order to complete.  I don't know what auto progress really means there.

**Comment by shijin-aws** [comment]
AWS CI failed due to ubertest failed to parse file

```
--------------------------------- Captured Out ---------------------------------

server_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.38.120 'timeout 1000 /bin/bash --login -c '"'"'FI_LOG_LEVEL=warn /home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10009-debug/install/fabtests/bin/fi_ubertest -x'"'"''

client_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.38.120 'timeout 1000 /bin/bash --login -c '"'"'FI_LOG_LEVEL=warn /home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10009-debug/install/fabtests/bin/fi_ubertest -u /home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10009-debug/install/fabtests/share/fabtests/test_configs/sockets/quick.test 172.31.38.120'"'"''
client_stdout:
[error] fabtests:ubertest/config.c:316: Unknown wait_obj
[error] fabtests:ubertest/config.c:568: Error parsing config!
[error] fabtests:ubertest/config.c:629: Unable to parse file

client returncode: 1
server_stdout:

server returncode: 124

```

**Comment by j-xiong** [comment]
@shijin-aws Thanks. Similarly, the failure in Intel CI is related to trying to run removed test (fi_poll).

File: include/rdma/fabric.h [hunk 2]
```diff
@@ -779,11 +779,14 @@ struct fi_param {
...
 
-/* Dummy definitions for removed flags/caps/types. For compiling old fabtests */
+/* Dummy definitions for removed flags/caps/types. For compiling old fabtests and MPICH */
 #define FI_VARIABLE_MSG		0ULL
 #define FI_NOTIFY_FLAGS_ONLY	0ULL
 #define FI_RESTRICTED_COMP	0ULL
 #define FI_EP_SOCK_STREAM	FI_EP_UNSPEC
+#define FI_ORDER_NONE		0ULL
+#define FI_MR_BASIC		(FI_MR_VIRT_ADDR | FI_MR_ALLOCATED | FI_MR_PROV_KEY)
+#define FI_MR_SCALABLE		0
```
Comment (shefty) [inline include/rdma/fabric.h:789]: Defining FI_MR_BASIC and FI_MR_SCALABLE allow the app to compile, but the behavior isn't the same.  FI_MR_BASIC, specifically will result is a broken app.  IMO, it's better to fail the compilation.

File: Makefile.am [hunk 3]
```diff
@@ -294,7 +293,7 @@ real_man_pages = \
         man/man7/fi_arch.7 \
         man/man7/fi_direct.7 \
         man/man7/fi_guide.7 \
-	man/man7/fi_hook.7 \
+	      man/man7/fi_hook.7 \
```
Comment (shefty) [inline Makefile.am:296]: tab got inserted here

File: include/rdma/fabric.h (hunk 2 above)
Comment (j-xiong) [reply include/rdma/fabric.h:799]: Keep these temporarily for current CI testing. Will remove before merge.

File: include/rdma/fabric.h [hunk 4]
```diff
@@ -336,6 +336,13 @@ enum {
 	FI_PROTO_CXI_RNR,
 };
 
+enum {
+	FI_TAG_BITS,
+	FI_TAG_HPC,
+	FI_TAG_AI,
```
Comment (shefty) [inline include/rdma/fabric.h:342]: Are we wanting HPC and AI or MPI and CCL?  The latter are closer aligned to what the definitions correspond to.

File: man/fi_endpoint.3.md [hunk 5]
```diff
@@ -782,40 +782,110 @@ A value of -1 guarantees ordering for any data size.
...
+  communication libraries and applications.  The AI format consists of a single
+  field: a payload identifier.  The identifier corresponds to the operation or
+  data being carried in the message payload.  For example, this field may be
+  used to identify whether a message is for point-to-point communication or
+  part of a collective operation, and in the latter case, the type of
+  collective operation.
+
+  The AI tag format does not require searching for matching receive
+  buffers, only directing the message to the correct virtual message queue
+  based on to the payload identifier.
```
Comment (shefty) [inline man/fi_endpoint.3.md:844]: extra 'to' in the text that isn't needed

File: man/fi_atomic.3.md [hunk 6]
```diff
@@ -228,6 +228,18 @@ provider implementation constraints.
   a complex number and the second representing the imaginary
   portion.
 
+*FI_FLOAT16*
+: 16-bit half precision floating point value (IEEE 754-2008).
+
+*FI_BFLOAT16*
+: 16-bit brain floating point value (IEEE 754-2008).
```
Comment (shefty) [inline man/fi_atomic.3.md:235]: brain floating point - I love IEEE

File: include/rdma/fabric.h (hunk 4 above)
Comment (j-xiong) [reply include/rdma/fabric.h:342]: MPI and CCL. Forgot to change that.

**Comment by miharulidze** [comment]
Dear all,

Are there any published schedule for the 2.0 API/library release and/or any a specific public mailing list or meeting where the ideas/release plans are discussed?

Thank you very much !

**Comment by shijin-aws** [comment]
For https://github.com/ofiwg/libfabric/pull/10009/commits/8f17f059accfd142f790c2aac75b258df1c4ea61, I think it doesn't have to be ofi 2.0 only, can we make a separate PR that can be merged to main now?

**Comment by j-xiong** [comment]
> Dear all,
> 
> Are there any published schedule for the 2.0 API/library release and/or any a specific public mailing list or meeting where the ideas/release plans are discussed?
> 
> Thank you very much !

There is a roadmap here:
https://github.com/ofiwg/libfabric/discussions/8049

Discussions are usually held on the bi-weekly ofiwg meeting on Tuesdays mornings: https://www.openfabrics.org/my-calendar/

**Comment by j-xiong** [comment]
@shijin-aws Yes, we can do that.

**Comment by j-xiong** [comment]
Replaced with #10279 and #10250 .

**Pull request closed without merging**